src/crypto/shelley/delegationUtils.js

```diff
@@ -0,0 +1,298 @@
+  if (asBase != null) {
+    if (
+      Buffer.from(await (await asBase.stake_cred()).to_bytes()).toString(
+        'hex',
+      ) === accountKeyString
+    ) {
+      return true
+    }
+  }
+  // TODO(v-almonacid): PointerAddress not yet implemented
+  // const asPointer = await PointerAddress.from_address(wasmAddr);
+  // if (asPointer != null) {
+  //   // TODO
+  // }
+  return acceptTypeMismatch
+}
+
+const filterAddressesByStakingKey = async (
+  stakingKey: StakeCredential,
+  utxos: $ReadOnlyArray<$ReadOnly<AddressedUtxo>>,
+  acceptTypeMismatch: boolean,
+): Promise<$ReadOnlyArray<$ReadOnly<AddressedUtxo>>> => {
+  const result = []
+  for (const utxo of utxos) {
+    if (
+      await addrContainsAccountKey(
+        utxo.receiver,
+        stakingKey,
+        acceptTypeMismatch,
+      )
+    ) {
+      result.push(utxo)
+    }
+  }
+  return result
+}
+
+/**
+ * Sending the transaction may affect the amount delegated in a few ways:
+ * 1) The transaction fee for the transaction
+ *  - may be paid with UTXO that either does or doesn't belong to our staking key.
+ * 2) The change for the transaction
+ *  - may get turned into a group address for our staking key
+ */
+const getDifferenceAfterTx = async (
+  utxoResponse: V4UnsignedTxAddressedUtxoResponse,
+  allUtxos: Array<AddressedUtxo>,
+  stakingKey: PublicKey,
+): Promise<BigNumber> => {
+  const stakeCredential = await StakeCredential.from_keyhash(stakingKey.hash())
+
+  let sumInForKey = new BigNumber(0)
+  {
+    // note senderUtxos.length is approximately 1
+    // since it's just to cover transaction fees
+    // so this for loop is faster than building a map
+    for (const senderUtxo of utxoResponse.senderUtxos) {
+      const match = allUtxos.find(
+        (utxo) =>
+          utxo.tx_hash === senderUtxo.tx_hash &&
+          utxo.tx_index === senderUtxo.tx_index,
+      )
+      if (match == null) {
+        throw new Error(
+          'getDifferenceAfterTx: utxo not found. Should not happen',
+        )
+      }
+      const address = match.receiver
+      if (await addrContainsAccountKey(address, stakeCredential, true)) {
+        sumInForKey = sumInForKey.plus(new BigNumber(senderUtxo.amount))
+      }
+    }
+  }
+
+  let sumOutForKey = new BigNumber(0)
+  {
+    const txBody = await utxoResponse.txBuilder.build()
+    const outputs = await txBody.outputs()
+    for (let i = 0; i < (await outputs.len()); i++) {
+      const output = await outputs.get(i)
+      const address = Buffer.from(
+        await (await output.address()).to_bytes(),
+      ).toString('hex')
+      if (await addrContainsAccountKey(address, stakeCredential, true)) {
+        const value = new BigNumber(await (await output.amount()).to_str())
+        sumOutForKey = sumOutForKey.plus(value)
+      }
+    }
+  }
+
+  return sumOutForKey.minus(sumInForKey)
+}
+
+export type CreateDelegationTxRequest = {|
+  absSlotNumber: BigNumber,
+  // computeRegistrationStatus: (void) => Promise<boolean>,
```

Comment:
One thing I still need to figure out is how to get and store certificate registration history. I think that I'll omit that part for now and just use the `accountState` endpoint to determine whether a given key has a registered certificate or not.

Reply:
>accountState
the account state endpoint doesn't tell you this FYI. The registration history endpoint (not deployed yet) and the tx history endpoint can tell you this however